### ISSUE TITLE
Homebrew support

### ISSUE THREAD

**Issue by radex** [krzysztofzablocki/Sourcery#15]
It would be nice to have Sourcery in Homebrew.

Since it's part of the dev tooling, and not the build itself (arguably), I'd prefer not to use CocoaPods for this, but would still be nice to have easy updates with a dependency manager.

**Comment by krzysztofzablocki** [comment]
In general I would say it's better if it's part of the build process, because it will prevent the bugs I mentioned in readme when people refactor code, but I agree it would be convienent to also be able to grab this via `brew`.

There issue here is that `brew` formulas don't work with apps, we'd have to use cask for apps. 
I had to make `Sourcery` a mac app because it needs to be able to have embedded swift frameworks and it doesn't work with Swift CLI.

There is a way to do it by manually decomposing application, an example of that is [here](https://github.com/AliSoftware/SwiftGen/blob/master/Rakefile).

It's definitely on the my todo list, but if anyone has time / experience with `brew` formulas and would like to create Rakefile that we could use that would be amazing :)

**Comment by vknabel** [comment]
There shouldn't be a problem with frameworks when using the SwiftPM. At least a CLI did work, when copying to another Mac.

**Comment by krzysztofzablocki** [comment]
Now that we have SPM support, maybe someone would like to contribute `brew` support? it shouldn't be too hard given spm builds CLI not an app. 

Right now I'm working on writing more examples, improving our parser speed, so not sure when I'll be free to work on this, so if anyone has some time that would be awesome thing to contribute as it would help people play with the tool or use daemon mode.

[We'd need to add something like this](https://github.com/AliSoftware/SwiftGen/blob/master/Rakefile#L273) to our Rakefile and just tweak the formula

**Comment by swizzlr** [comment]
Dropping some notes in here on my initial – frustrating – attempt:

1. I got it working!
2. Homebrew doesn't yet seem to support swift build – it can't find clang or something. No idea why.
2. Right now my Formula.rb is:

```ruby
# Documentation: http://docs.brew.sh/Formula-Cookbook.html
#                http://www.rubydoc.info/github/Homebrew/brew/master/Formula
# PLEASE REMOVE ALL GENERATED COMMENTS BEFORE SUBMITTING YOUR PULL REQUEST!

class Sourcery < Formula
  desc "Meta-programming for Swift, stop writing boilerplate code."
  homepage "https://github.com/krzysztofzablocki/Sourcery"
  url "https://github.com/krzysztofzablocki/Sourcery/archive/0.5.3.tar.gz"
  sha256 "ad34e2146e73ce81894ac1dd9130fb2fd12c8213789aef7c5cde9580f813c823"

  depends_on :xcode => ["8.0", :build]

  def install
    # ENV.deparallelize  # if your formula fails when building in parallel

    # Remove unrecognized options if warned by configure
    rake "build"
    bin.install Dir["build/*"]
  end

  test do
    # `test do` will create, run in and delete a temporary directory.
    #
    # This test will fail and we won't accept that! It's enough to just replace
    # "false" with the main program this formula installs, but it'd be nice if you
    # were more thorough. Run the test with `brew test Sourcery`. Options passed
    # to `brew install` such as `--HEAD` also need to be provided to `brew test`.
    #
    # The installed folder is not in the path, so use the entire path to any
    # executables being tested: `system "#{bin}/program", "do", "something"`.
    system "#{bin}/sourcery", "--version"
  end
end

```

**Comment by krzysztofzablocki** [comment]
Will homebrew accept a formula that creates an app? I thought that's going to be rejected, maybe @vknabel has idea how to get swift pm to work with formulas?

**Comment by radex** [comment]
My understanding is that you can't distribute pre-compiled apps with Homebrew (Caskroom is for that), but you can have a formula that builds its own binaries.

**Comment by swizzlr** [comment]
It's still compiling from source, I guess. But it would be good to see about SPM builds in Himebrewz

**Comment by vknabel** [comment]
@swizzlr I could fix the clang error thanks to [jakeheis/homebrew-repo/flock.rb](https://github.com/jakeheis/homebrew-repo/blob/master/flock.rb), but I am not sure what `unset CC` means. But it seems to work. Additionally some indirect dependencies are linked dynamically (e.g. `libCYaml.dylib`), so some libraries need to be linked.


@radex @krzysztofzablocki @swizzlr For me the following formula works! 🎉

```ruby
# Documentation: http://docs.brew.sh/Formula-Cookbook.html
#                http://www.rubydoc.info/github/Homebrew/brew/master/Formula
# PLEASE REMOVE ALL GENERATED COMMENTS BEFORE SUBMITTING YOUR PULL REQUEST!

class Sourcery < Formula
  desc "Meta-programming for Swift, stop writing boilerplate code."
  homepage "https://github.com/krzysztofzablocki/Sourcery"
  url "https://github.com/krzysztofzablocki/Sourcery/archive/0.5.3.tar.gz"
  sha256 "ad34e2146e73ce81894ac1dd9130fb2fd12c8213789aef7c5cde9580f813c823"

  depends_on :xcode => ["8.0", :build]

  def install
    # ENV.deparallelize  # if your formula fails when building in parallel

    # Remove unrecognized options if warned by configure
    system "unset CC; swift build -c release"
    bin.install ".build/release/sourcery"
    lib.install Dir[".build/release/*.dylib"]
  end

  test do
    # `test do` will create, run in and delete a temporary directory.
    #
    # This test will fail and we won't accept that! It's enough to just replace
    # "false" with the main program this formula installs, but it'd be nice if you
    # were more thorough. Run the test with `brew test Sourcery`. Options passed
    # to `brew install` such as `--HEAD` also need to be provided to `brew test`.
    #
    # The installed folder is not in the path, so use the entire path to any
    # executables being tested: `system "#{bin}/program", "do", "something"`.
    system "#{bin}/sourcery", "--version"
  end
end
```

**Comment by swizzlr** [comment]
Propose it dude!

**Comment by vknabel** [comment]
@radex @krzysztofzablocki @swizzlr You can find the PR here: Homebrew/homebrew-core#9191

**Comment by swizzlr** [comment]
This issue can now be closed by the fabulous work of @vknabel!

**Comment by krzysztofzablocki** [comment]
@vknabel Thank you so much!

@swizzlr we need to add readme entry for it, want to send PR ? :) otherwise I'll try to do it after work

**Comment by swizzlr** [comment]
@vknabel how did you generate the brew formula?

@krzysztofzablocki will do

**Comment by vknabel** [comment]
@swizzlr I started with your Formula, used `swift build -c release` instead of `rake build` and incrementally got everything to work through trial and error. Then I added some tests by hand (which I needed to disable)

**Comment by krzysztofzablocki** [comment]
Any update @swizzlr ?

**Comment by jpsim** [comment]
Just an FYI, I was looking at Sourcery's (unofficial?) Homebrew formula as I was entertaining the idea of using the Swift Package Manager to perform the installation, and noticed a few things that I'll share here:

The current Homebrew formula [here](https://github.com/Homebrew/homebrew-core/blob/master/Formula/sourcery.rb) _requires_ that the user have the exact same version of Xcode as Homebrew's Jenkins CI machines located exactly at the same location (`/Applications/Xcode.app`).

That's because the homebrew formula uses Swift Package Manager to build the bottle, which adds an `LC_RPATH` command to the binary with an absolute path to the SDK where the Swift runtime dynamic libraries can be found (e.g. `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx`). So if you've installed Xcode anywhere else, or if the version of Xcode you have installed at that location isn't the same version as was used on Homebrew's Jenkins machines, you'll have to build from source using a command like `brew install sourcery --HEAD`.

**Comment by jpsim** [comment]
Here's my attempt at fixing this: Homebrew/homebrew-core#12203

Based on this Twitter thread: https://twitter.com/simjp/status/850393838753464321

**Comment by ilovezfs** [comment]
It looks like the attempt was successful on El Capitan but failed on Sierra, I believe because our Sierra nodes are running Xcode 8.3.1, and there's some kind of incompatibility between sourcery and Xcode 8.3.x, which leads to
```
==> swift build -c release -Xswiftc -static-stdlib
swift-build: error: the package has an unsupported layout, modulemap (/private/tmp/sourcery-20170412-49020-7oic81/Sourcery-0.5.9/.build/checkouts/SourceKitten.git-6484296299232452758/Source/SourceKittenFramework/clang-c/module.modulemap) is not allowed to be mixed with sources
fix: move the modulemap inside include directory
Fetching https://github.com/kylef/Stencil.git
Fetching https://github.com/kylef/Commander.git
Fetching https://github.com/kylef/PathKit.git
Fetching https://github.com/jpsim/SourceKitten.git
Fetching https://github.com/vknabel/SwiftTryCatch.git
Fetching https://github.com/IBM-Swift/CommonCrypto.git
Fetching https://github.com/SwiftGen/StencilSwiftKit.git
Fetching https://github.com/kylef/Spectre.git
Fetching https://github.com/Carthage/Commandant.git
Fetching https://github.com/drmohundro/SWXMLHash.git
Fetching https://github.com/jpsim/Yams.git
Fetching https://github.com/norio-nomura/Clang_C.git
Fetching https://github.com/norio-nomura/SourceKit.git
Fetching https://github.com/antitypical/Result.git
Cloning https://github.com/jpsim/SourceKitten.git
Resolving https://github.com/jpsim/SourceKitten.git at 0.15.3
Cloning https://github.com/jpsim/Yams.git
Resolving https://github.com/jpsim/Yams.git at 0.1.2
Cloning https://github.com/antitypical/Result.git
Resolving https://github.com/antitypical/Result.git at 3.2.1
Cloning https://github.com/kylef/Stencil.git
Resolving https://github.com/kylef/Stencil.git at 0.8.0
Cloning https://github.com/IBM-Swift/CommonCrypto.git
Resolving https://github.com/IBM-Swift/CommonCrypto.git at 0.1.4
Cloning https://github.com/norio-nomura/Clang_C.git
Resolving https://github.com/norio-nomura/Clang_C.git at 1.0.2
Cloning https://github.com/kylef/Commander.git
Resolving https://github.com/kylef/Commander.git at 0.6.1
Cloning https://github.com/kylef/PathKit.git
Resolving https://github.com/kylef/PathKit.git at 0.8.0
Cloning https://github.com/norio-nomura/SourceKit.git
Resolving https://github.com/norio-nomura/SourceKit.git at 1.0.1
Cloning https://github.com/SwiftGen/StencilSwiftKit.git
Resolving https://github.com/SwiftGen/StencilSwiftKit.git at 1.0.1
Cloning https://github.com/vknabel/SwiftTryCatch.git
Resolving https://github.com/vknabel/SwiftTryCatch.git at 1.1.0
Cloning https://github.com/kylef/Spectre.git
Resolving https://github.com/kylef/Spectre.git at 0.7.2
Cloning https://github.com/drmohundro/SWXMLHash.git
Resolving https://github.com/drmohundro/SWXMLHash.git at 3.0.3
Cloning https://github.com/Carthage/Commandant.git
Resolving https://github.com/Carthage/Commandant.git at 0.11.3
```

**Comment by ilovezfs** [comment]
@krzysztofzablocki has the above been addressed in HEAD yet, or should we open a separate issue for this?

**Comment by jpsim** [comment]
Yes it's addressed in HEAD, see https://github.com/Homebrew/homebrew-core/pull/12203#issuecomment-293667340 for details.

**Comment by orta** [comment]
in the CP Mac app, we automate generating the PR for [each new version](https://github.com/CocoaPods/CocoaPods-app/blob/master/Rakefile#L1217-L1259)

**Comment by Antondomashnev** [comment]
@ilovezfs I've been working on the task for Sourcery release process to submit a Homebrew PR and since I've never worked on formulas before I have one question:
In the [Homebrew CONTRIBUTING](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md) is says that to bump the formula version the `brew bump-formula-pr --strict` should be used. Can I use instead of it the following commands:
```sh
brew audit --strict --online #{formula}
brew upgrade #{formula}
brew test #{formula}
```

**Comment by ilovezfs** [comment]
No, it will only run `brew audit --strict foo` and then open the pull request.

Before doing that you do the steps you posted above locally, and then revert the formula to its original state before running the bump-formula-pr command.

Note that audit should be run after install not before as mentioned in the PR template.

**Comment by Antondomashnev** [comment]
@ilovezfs thanks for the tips. Can I open PR through the API post request like @orta mentioned in [CocoaPods app](https://github.com/CocoaPods/CocoaPods-app/blob/master/Rakefile#L1256-L1258)?

> Note that audit should be run after install not before as mentioned in the PR template.

Will do so, thanks 👍

**Comment by ilovezfs** [comment]
@Antondomashnev probably, but bump-formula-pr is easiest because it handles thinks like removing revisions from the formula.

**Comment by krzysztofzablocki** [comment]
is there anything left or do we close this? @Antondomashnev @ilyapuchka

**Comment by Antondomashnev** [comment]
@krzysztofzablocki the only missing thing is a real test on the new release step.

**Comment by Lutzifer** [comment]
seems to work, @Antondomashnev  ?

**Comment by Antondomashnev** [comment]
Well, during the last release I had to do it manually and it didn't work in the `rake release:new`. But the error was related to my laptop's configuration and I don't think it's a generic error.

**Comment by krzysztofzablocki** [comment]
Ok, I'm closing this then, and we'll see if something has to be tweaked on next release
